File: src/main/java/discourje/core/validation/rules/ClosedChannelMustBeUsedInProtocol.java
package discourje.core.validation.rules;

import discourje.core.validation.Rule;
import discourje.core.validation.formulas.CtlFormula;
import static discourje.core.validation.formulas.CtlFormulas.EF;
import static discourje.core.validation.formulas.CtlFormulas.close;
import static discourje.core.validation.formulas.CtlFormulas.implies;
import static discourje.core.validation.formulas.CtlFormulas.msg;

public class ClosedChannelMustBeUsedInProtocol extends Rule {

    @Override
    public String createErrorDescription(String r1, String r2) {
        return String.format("A channel from %s to %s is closed, but this channel is never used in the protocol.", r1, r2);
    }

    @Override
    public CtlFormula createCtlFormula(String r1, String r2) {
        return implies(EF(close(r1, r2)), EF(msg(r1, r2)));
    }
}

File: src/test/clojure/discourje/core/ctl/closed_channel_must_be_used_in_protocol.clj
(ns discourje.core.validation.closed-channel-must-be-used-in-protocol
  (:require [discourje.core.spec :as s]))

(s/defrole ::a)
(s/defrole ::b)
(s/defrole ::c)
(s/defrole ::d)

;; All closed channels are also used
(s/defsession ::protocol-trivial-correct []
              [
               (s/--> ::a ::b)
               (s/--> ::b ::a)
               (s/close ::a ::b)
               (s/close ::b ::a)
               ])

;; Not all closed channels are also used
(s/defsession ::protocol-trivial-incorrect []
              [
               (s/--> ::a ::b)
               (s/close ::a ::b)
               (s/close ::b ::a)
               ])

;; Channels are closed on both paths, but used on only one
(s/defsession ::protocol-non-trivial-correct []
              (s/alt
                [(s/--> ::a ::b)
                 (s/close ::a ::b)
                 (s/close ::b ::a)
                 ]
                [(s/--> ::b ::a)
                 (s/close ::b ::a)
                 (s/close ::a ::b)
                 ]
                ))

;; channel b c is closed, but not used, not even on other paths
(s/defsession ::protocol-non-trivial-incorrect []
              (s/alt
                [(s/--> ::a ::b)
                 (s/close ::a ::b)
                 (s/close ::b ::c)
                 ]
                [(s/--> ::b ::a)
                 (s/close ::a ::b)
                 (s/close ::b ::a)
                 ]
                ))
